File: practices/chapter7_heap/python/ch_7_heap_tree.py
class Heap_tree():
  def __init__(self):
    self.heap = []
    self.size = 0
    pass

  def get_child_left_node_index(self, index):
    return index * 2 + 1
  def get_child_right_node_index(self, index):
    return index * 2 + 2

  def get_parent_node_index(self, index):
    return (index - 1) // 2
  def get_child_node_min_index(self, index):
    # 找該節點的右邊子節點(等同node.right)
    # 如果找不到，則給該節點的左子節點
    child_left_node_index = self.get_child_left_node_index(index)
    child_right_node_index = self.get_child_right_node_index(index)

    if child_right_node_index >= self.size:
      return child_left_node_index # 左節點
      pass
    else:
      if self.heap[child_left_node_index] < self.heap[child_right_node_index]:
        return child_left_node_index
        pass
      return child_right_node_index

  def data_down_change(self, index=0):
    _index = index
    # 對調節點
    while self.get_child_right_node_index(_index) <= self.size:
      min_index = self.get_child_node_min_index(_index)
      # 如果該節點比該節點下的子節點最小還"大"
      if self.heap[_index] > self.heap[min_index]:
        # 對調
        self.heap[_index], self.heap[min_index] = self.heap[min_index], self.heap[_index]
      # 繼續往下找
      _index = min_index
    pass

  def data_up_change(self, index=0):
    _index = index
    while self.get_parent_node_index(_index) >= 0:
      parent_node_index = self.get_parent_node_index(_index)
      # 如果該節點比父節點還小
      if self.heap[_index] < self.heap[parent_node_index]:
        # 對調節點
        self.heap[_index], self.heap[parent_node_index] = self.heap[parent_node_index], self.heap[_index]
      # 繼續往上找
      _index = parent_node_index
    pass

  def create_tree(self, list):
    self.heap = list
    self.size = len(list)
    index = (len(list) // 2) - 1 # 找到塞滿子節點的那個index
    while (index >= 0):
      self.data_down_change(index)
      index = index - 1
    pass

  def get_min_node(self):
    min_node = self.heap[0]
    self.size -= 1
    self.heap[0] = self.heap[self.size] # 用最尾端的節點塞到初始
    self.heap.pop() # 把原本尾端的拿掉
    self.data_down_change(0)
    return min_node
    pass

  def append(self, data):
    self.heap.append(data) # 塞在最後
    self.size += 1
    index = self.size - 1
    while(index >= 0):
      self.data_up_change(index)
      index = (index - 1) // 2
    return self.heap


dataArr = [10, 21, 5, 9, 13, 28, 3]
heap_tree = Heap_tree()
heap_tree.create_tree(dataArr)

def heap_tree_sample():
  print('heap tree: ', heap_tree.heap)

def heap_tree_get_min_sample():
  tree = heap_tree
  print('Get min node: ', heap_tree.get_min_node())
  print('Min node removed: ', heap_tree.heap)

  print('Get min node: ', heap_tree.get_min_node())
  print('Min node removed(2): ', heap_tree.heap)

def heap_tree_append_sample():
  tree = heap_tree
  for i in range(3):
    inser_val = eval(input('Insert value: '))
    print(tree.append(inser_val))
    pass
  # print('Insert value: 2', tree.append(2))
  # print('Inserted result: ', tree.heap)

if __name__ == "__main__":
    # heap_tree_sample()
    # heap_tree_get_min_sample()
    heap_tree_append_sample()

File: practices/chapter7_heap/js/ch_7_heap_tree.js
const {
  rl,
} = require('../../lib/cmdReadline')

class HeapTreeIndexHadlers {
  static getChildLeftNodeIndex(index=0) {
    return index * 2 + 1
  }

  static getChildRightNodeIndex(index=0) {
    return index * 2 + 2
  }

  static getParentNodeIndex(index=0) {
    return Math.floor((index - 1) / 2)
  }

  static getChildNodeMinIndex(heapTree=new HeapTree(), index=0) {
    const childLeftNodeIndex = this.getChildLeftNodeIndex(index)
    const childRightNodeIndex = this.getChildRightNodeIndex(index)

    // 如果沒有右邊子節點(右邊子節點的index超過size)
    if (childRightNodeIndex >= heapTree.size) {
      return childLeftNodeIndex
    } else {
      if(heapTree.heap[childLeftNodeIndex] < heapTree.heap[childRightNodeIndex]) {
        return childLeftNodeIndex
      }
      return childRightNodeIndex
    }
  }
}

class HeapTree {
  constructor() {
    this.heap = []
    this.size = this.heap.length
  }

  dataDownChange(index) {
    let _idx = index
    while(HeapTreeIndexHadlers.getChildRightNodeIndex(_idx) <= this.size) {
      let childMinNodeIndex = HeapTreeIndexHadlers.getChildNodeMinIndex(this, _idx)
      // console.log(childMinNodeIndex)
      if(this.heap[_idx] > this.heap[childMinNodeIndex]) {
        // 對調
        [this.heap[_idx], this.heap[childMinNodeIndex]] = 
        [this.heap[childMinNodeIndex], this.heap[_idx]]
      }
      _idx = childMinNodeIndex
    }
    // console.log(this.heap)
  }

  dataUpChange(index=0) {
    let idx = index
    while(HeapTreeIndexHadlers.getParentNodeIndex(idx) >= 0) {
      const parentNodeIdx = HeapTreeIndexHadlers.getParentNodeIndex(idx)
      // 如果該節點比父節點還小則對調
      if(this.heap[idx] < this.heap[parentNodeIdx]) {
        [this.heap[parentNodeIdx], this.heap[idx]] = 
        [this.heap[idx], this.heap[parentNodeIdx]]
      }
      idx = parentNodeIdx
    }
  }

  create(dataList=[]) {
    this.heap = dataList
    this.size = dataList.length
    let index = Math.floor(this.size / 2) - 1
    while(index >= 0) {
      this.dataDownChange(index)
      index -= 1
    }
  }

  // 塞新的值
  append(data) {
    this.heap.push(data)
    this.size += 1
    let idx = this.size - 1
    while(idx >= 0) {
      this.dataUpChange(idx)
      idx = HeapTreeIndexHadlers.getParentNodeIndex(idx)
    }
    return this.heap
  }

  // 取得最小值
  pop() {
    const min = this.heap[0]
    this.size -= 1
    this.heap[0] = this.heap[this.size]
    this.heap.pop()
    this.dataDownChange(0)
    return min
  }
}

const dataArr = [10, 21, 5, 9, 13, 28, 3]
const heapTree = new HeapTree()
heapTree.create(dataArr)

function heapTreeSample() {
  console.log('Heap tree: ', heapTree.heap)
}

function heapTreeInsertSample() {
  rl.question('Insert value: ', ans => {
    heapTree.append(Number(ans))
    console.log('Heap tree: ', heapTree.heap)
    rl.close()
  })
}

function heapTreePopSample() {
  console.log('Heap tree: ', heapTree.heap)
  
  console.log('Heap tree pop: ', heapTree.pop())
  console.log('Heap tree poped(first): ', heapTree.heap)

  console.log('Heap tree pop: ', heapTree.pop())
  console.log('Heap tree poped(second): ', heapTree.heap)
  
  console.log('Heap tree pop: ', heapTree.pop())
  console.log('Heap tree poped(third): ', heapTree.heap)
  rl.close()
}

function main() {
  // heapTreeSample()
  // heapTreeInsertSample()
  heapTreePopSample()
}

main()

